src/main/java/com/wavefront/SpanGenerator.java

```diff
@@ -94,28 +94,97 @@ public SpanQueue generate(GeneratorConfig config) {
     tags.add(new Pair<>("host", "ip-10.20.30.40"));
 
     if (errorRate > 0) {
-      if (randGenerator.nextInt(100) < errorRate) {
+      if (RANDOM.nextInt(100) < errorRate) {
         tags.add(new Pair<>("error", "true"));
       }
     }
 
     return tags;
   }
+
+  /**
+   * Normalize all distributions. ie trace type, spans counts and so on.
+   */
+  private void normalizeDistributions(List<TraceTypePattern> traceTypePatterns) {
+    // trace types distribution
+    double sum = traceTypePatterns.stream().mapToDouble(d -> d.tracePercentage).sum();
+    traceTypePatterns.forEach(d ->
+        d.tracePercentage = (int) Math.round(d.tracePercentage * getNormalizationRatio(sum)));
+
+    // spans distribution
+    traceTypePatterns.forEach(traceTypePattern -> {
+      double spansSum = traceTypePattern.spansDistributions.stream().mapToDouble(d -> d.percentage).sum();
+      traceTypePattern.spansDistributions.forEach(d ->
+          d.percentage = (int) Math.round(d.percentage * getNormalizationRatio(spansSum)));
+    });
+  }
+
+  /**
+   * Get normalization ratio for given sum of percentages.
+   */
+  private double getNormalizationRatio(double percentsSum) {
+    double ratio = 1;
+    if (Double.compare(percentsSum, HUNDRED_PERCENT) != 0) {
+      logger.warning("Distributions summary percentage must be 100. " +
+          "Normalizing in range 0-100");
+      ratio = (double) HUNDRED_PERCENT / percentsSum;
+    }
+    return ratio;
+  }
+
+  /**
+   * Get next span distribution for the given trace type.
+   */
+  private Distribution getNextSpanDistribution(TraceTypePattern traceTypePattern) {
+    List<Integer> percentages = traceTypePattern.spansDistributions.stream().map(distribution ->
+        distribution.percentage).collect(Collectors.toList());
+    return traceTypePattern.spansDistributions.get(getIndexOfNextItem(percentages));
+  }
+
+  /**
+   * Get nex trace type for generation based on trace type distributions.
+   *
+   * @param traceTypePatterns List of trace types to be generated.
+   */
+  private TraceTypePattern getNextTraceType(List<TraceTypePattern> traceTypePatterns) {
+    List<Integer> percentages = traceTypePatterns.stream().
+        map(traceTypePattern -> traceTypePattern.tracePercentage).collect(Collectors.toList());
+    return traceTypePatterns.get(getIndexOfNextItem(percentages));
+  }
+
+  /**
+   * Generic helper method which calculate next item based on distribution percentages using
+   * randomization.
+   *
+   * @param percentages List of distribution percentages.
+   * @return Return an index of the next item.
+   */
+  private int getIndexOfNextItem(List<Integer> percentages) {
+    int randomPercent = RANDOM.nextInt(HUNDRED_PERCENT) + 1;
+    int left = 0;
+    for (int i = 0; i < percentages.size(); i++) {
+      if (randomPercent > left && randomPercent <= percentages.get(i) + left) {
+        return i;
+      } else {
+        left += percentages.get(i);
+      }
+    }
```

Comment:
Usually, I avoid the call of the same method twice and prefer to save the result in a variable and use it. But maybe Java compiler does this optimization.